### PR TITLE
8312434: SPECjvm2008/xml.transform with CDS fails with "can't seal package nu.xom"

### PR DESCRIPTION
**Pull request by calvinccheung**
Problem:
A jar file containing classes in a sealed package and is signed with a "disabled" algorithm as indicated in the  `jdk.jar.disabledAlgorithms` security property. Some of the classes are stored in a CDS archive. During runtime, if a class is loaded from the archive followed by loading a class from the jar file, resulting in sealing violation.

Cause:
During dump time, CDS considers the jar file as signed although it should be treated as unsigned due to the algorithm used is in the "disabled" list. Currently, CDS doesn't store the manifest of a signed jar in the archive since CDS doesn't support signed classes. During runtime, since there's no manifest info, when a class is loaded from the archive, a package entry is created without sealing information. When a subsequent class from the same package is loaded from the jar, an attempt to create a package entry with sealing information would fail.

Fix:
It is difficult in the hotspot code to determine if an algorithm for signing the jar is considered "disabled". The fix is to always store the jar manifest in the CDS archive.

Testing:
Passed tiers 1 - 4 (including the new test).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312434](https://bugs.openjdk.org/browse/JDK-8312434): SPECjvm2008/xml.transform with CDS fails with "can't seal package nu.xom" (**Bug** - P4)

### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15339/head:pull/15339` \
`$ git checkout pull/15339`

Update a local copy of the PR: \
`$ git checkout pull/15339` \
`$ git pull https://git.openjdk.org/jdk.git pull/15339/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15339`

View PR using the GUI difftool: \
`$ git pr show -t 15339`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15339.diff">https://git.openjdk.org/jdk/pull/15339.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15339#issuecomment-1683362069)

### CODE REVIEW

**Review by iklam — COMMENTED** [review]
Looks good to me with a small nit for the test case.

Also, can we remove the `SharedClassPathEntry::is_signed()` API as well?

**Comment by calvinccheung** [comment]
> Looks good to me with a small nit for the test case.
> 
> Also, can we remove the `SharedClassPathEntry::is_signed()` API as well?

Yes, I will also remove the `signed_jar_entry` enum.

File: test/hotspot/jtreg/runtime/cds/appcds/JarBuilder.java [hunk 1]
```diff
@@ -259,21 +259,36 @@ public static void compileModule(Path src,
         }
     }
 
+    static final String keyTool = JDKToolFinder.getJDKTool("keytool");
+    static final String jarSigner = JDKToolFinder.getJDKTool("jarsigner");
 
-    public static void signJar() throws Exception {
-        String keyTool = JDKToolFinder.getJDKTool("keytool");
-        String jarSigner = JDKToolFinder.getJDKTool("jarsigner");
+    public static void signJarWithDisabledAlg(String jarName) throws Exception {
```
[the remark this answers is not in the document]
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/JarBuilder.java:265]: Fixed. Also removed an extra import from the test.

**Review by iklam — APPROVED** [review]
LGTM

**Review by matias9927 — APPROVED** [review]
LGTM!

**Pull request closed without merging**